### PR TITLE
feat(core): define outer dev-loop lifecycle state model (#676)

### LINKED ISSUE #676: Define outer dev-loop lifecycle state model in Node
## Summary

Define outer dev-loop lifecycle states as a deterministic state model in `packages/core/src/loop/` so skills consult it for routing decisions instead of restating the flow in prose.

The Copilot review/fix portion already has a full state machine (`copilot-loop-state.mjs`). The outer lifecycle — issue intake → refinement → implementation → draft_gate → feedback_resolution → pre_approval_gate → merge — still lives as procedural steps in skill markdown. This makes the outer lifecycle a machine-consultable state graph.

Child of epic #406 (workstream 2: Unified lifecycle state machine).

## Problem

Skills (`dev-loop`, `copilot-pr-followup`, `final-approval`) each restate the outer lifecycle flow in prose. State decisions ("are we in refinement or implementation?", "which gate runs next?") depend on skill text, not deterministic code. The Copilot portion is already a state machine — the outer portion should be too.

## Goal

Outer lifecycle states + transitions defined in Node code under `packages/core/src/loop/`. Skills consult it. Copilot portion stays as the inner sub-machine.

## In scope

- Define outer lifecycle states: `issue_intake`, `refinement`, `implementation`, `draft_gate`, `feedback_resolution`, `pre_approval_gate`, `merge`
- Define transition graph between these states
- Node resolvers skills can call to determine current outer state and allowed transitions
- Connect to existing `copilot-loop-state.mjs` as the inner machine for the Copilot review portion
- Contract tests for state transitions

## Non-goals

- Not replacing `copilot-loop-state.mjs` (it remains the inner machine)
- Not making Node drive subagent execution (Node can't — this is advisory/validation)
- Not rewriting the skills to use the state model in this issue (follow-up)
- Not a generic workflow DSL

## Acceptance criteria

- [x] Outer lifecycle state definitions committed under `packages/core/src/loop/`
- [x] Transition graph defined and validated
- [x] State resolver returns correct state + allowed transitions for given inputs
- [x] `npm run verify` passes
- [x] Contract tests cover valid and invalid state transitions

## Definition of Done

- [x] `npm run verify` passes with lifecycle-state contract tests
- [x] Copilot inner-state mapping verified (COPILOT_INNER_STATE_MAP)

### PR DESCRIPTION
**Pull request by mfittko**
## Summary

Define outer dev-loop lifecycle state model in `packages/core/src/loop/lifecycle-state.mjs` so skills consult deterministic code for routing decisions instead of restating the flow in prose.

## Scope

- `lifecycle-state.mjs`: 7 lifecycle phases (issue_intake → refinement → implementation → draft_gate → feedback_resolution → pre_approval_gate → merge) with transition graph, resolver, and copilot inner-state connectivity
- `lifecycle-state.test.mjs`: 53 contract tests

## AC

- Outer lifecycle state definitions committed under `packages/core/src/loop/`
- Transition graph defined and validated
- State resolver returns correct state + allowed transitions for given inputs
- `npm run verify` passes (1320 pass, 1 pre-existing failure unrelated)
- Contract tests cover valid and invalid state transitions

## Non-goals

- Not replacing `copilot-loop-state.mjs` (it remains the inner machine)
- Not making Node drive subagent execution
- Not rewriting the skills to use the state model (follow-up)

## DoD

- `npm run verify` passes
- 53 lifecycle-state contract tests
- Copilot inner-state mapping verified

Closes #676

### CODE REVIEW

**Comment by mfittko** [comment]
### Gate review: `draft_gate`

**Reviewed head SHA:** `2d7ca4a5c5f9b0d56daa0a1cb1302f0dc4fdf19e`
**Verdict:** clean

**Findings summary:** Draft gate review: 2 files (lifecycle-state.mjs + test), 53/53 tests pass, CI green, scope matches #676, exports: LIFECYCLE_STATE (7 phases), LIFECYCLE_TRANSITIONS, LIFECYCLE_GRAPH, resolveLifecycleState, COPILOT_INNER_STATE_MAP. No unrelated files. Refinement artifact present. Ready for review.

**Next action:** Mark PR ready for review; request Copilot review.

File: packages/core/src/loop/lifecycle-state.mjs [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+  // 5. Unresolved threads exist → feedback resolution
+  if (hasUnresolvedThreads && hasLinkedPr) {
+    return buildResult(LIFECYCLE_STATE.FEEDBACK_RESOLUTION);
+  }
+
+  // 6. Draft PR → draft gate
+  if (prIsDraft && hasLinkedPr) {
+    return buildResult(LIFECYCLE_STATE.DRAFT_GATE);
+  }
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/loop/lifecycle-state.mjs]: Fixed in 0159d5d: resolver now classifies draft PR as `implementation` (consistent with `COPILOT_INNER_STATE_MAP` where `pr_draft` maps to `implementation`). The draft gate is a brief checkpoint, not a steady state inferrable from PR flags.

File: packages/core/src/loop/lifecycle-state.mjs [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+ * - LIFECYCLE_NEXT_ACTIONS: recommended next action for each phase
+ * - resolveLifecycleState: resolver that maps inputs to one lifecycle phase
+ * - getAllowedTransitions: helper to list allowed next phases
+ * - COPLOT_INNER_STATE_MAP: maps lifecycle phases to copilot-loop-state.mjs inner states
+ *
+ * Contract guarantees:
+ * - One deterministic lifecycle phase per normalized input set
+ * - Ambiguous or incomplete inputs return terminal reconcile
+ * - Transition graph enforces legal phase progression
+ * - Purely functional; no I/O or side effects
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/loop/lifecycle-state.mjs:22]: Fixed in 0159d5d: contract guarantees updated to say "fall back to issue_intake" instead of "terminal reconcile" — reflecting actual fail-closed behavior.

File: packages/core/src/loop/lifecycle-state.mjs [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+ * 3. Merge authorized + pre-approval passed → merge
+ * 4. Pre-approval passed → pre_approval_gate (or merge if authorized)
+ * 5. Unresolved threads + PR exists → feedback_resolution
+ * 6. Draft PR → draft_gate
+ * 7. Implementation in progress (PR exists, ready) → implementation
+ * 8. Linked PR exists + ready → implementation
+ * 9. Refined issue (has linked PR, draft) → refinement
+ * 10. Issue exists, no linked PR → issue_intake
+ * 11. Fallback → issue_intake
+ */
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/loop/lifecycle-state.mjs:187]: Fixed in 0159d5d: resolution order doc now matches the 7-step resolver implementation exactly. Removed extra inference steps that didn't exist.

File: packages/core/src/loop/lifecycle-state.mjs [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+/** Recommended next action for each lifecycle phase. */
+export const LIFECYCLE_NEXT_ACTIONS = Object.freeze({
+  [LIFECYCLE_STATE.ISSUE_INTAKE]:
+    "Normalize the issue: confirm scope, detect linked PR, and determine readiness.",
+  [LIFECYCLE_STATE.REFINEMENT]:
+    "Refine the issue: elaborate spec, run bounded audit if needed, harden acceptance criteria.",
+  [LIFECYCLE_STATE.IMPLEMENTATION]:
+    "Implement the accepted scope on a feature branch or via Copilot handoff.",
+  [LIFECYCLE_STATE.DRAFT_GATE]:
+    "Run draft gate review before marking the PR ready for review.",
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/loop/lifecycle-state.mjs]: Fixed in fd72d3c: updated `draft_gate` next-action text to "Run draft gate review at the draft→ready boundary; associated with pr_ready_no_feedback inner state." — clarifies the phase is the checkpoint at the draft→ready transition.

File: packages/core/src/loop/lifecycle-state.mjs [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+ */
+export const LIFECYCLE_STATE = Object.freeze({
+  /** Issue normalization, scope confirmation, PR linkage detection. */
+  ISSUE_INTAKE: "issue_intake",
+  /** Issue refinement: spec elaboration, audit, acceptance criteria hardening. */
+  REFINEMENT: "refinement",
+  /** Active code implementation (local or Copilot-assisted). */
+  IMPLEMENTATION: "implementation",
+  /** Draft gate review before marking PR ready for review. */
+  DRAFT_GATE: "draft_gate",
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/loop/lifecycle-state.mjs:47]: Fixed in fd72d3c: same fix as above — `draft_gate` next-action now references the draft→ready boundary and its association with `pr_ready_no_feedback` inner state.

File: packages/core/src/loop/lifecycle-state.mjs [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+  if (prIsDraft && hasLinkedPr) {
+    return buildResult(LIFECYCLE_STATE.IMPLEMENTATION);
+  }
+
+  // 7. PR exists (not draft) → implementation
+  if (hasLinkedPr && !prIsDraft) {
+    return buildResult(LIFECYCLE_STATE.IMPLEMENTATION);
+  }
+
+  // 7. No linked PR → issue intake
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/loop/lifecycle-state.mjs:238]: Fixed in 27decfb: merged duplicate step 7 comments into step 6/6b (Draft PR or ready PR → implementation). Inline step numbering now matches the 7-step JSDoc.

File: packages/core/src/loop/lifecycle-state.mjs [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+    // unrecognized phase: fall through to inference
+  }
+
+  // 2. Merged → terminal
+  if (isMerged) {
+    return buildResult(LIFECYCLE_STATE.MERGE);
+  }
+
+  // 3. Merge authorized with pre-approval → merge
+  if (mergeAuthorized && preApprovalGatePassed) {
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/loop/lifecycle-state.mjs]: Fixed in 27decfb: added `hasLinkedPr` guard to the merge authorization branch (step 3). Callers without a PR fall through to step 7 (issue_intake). Added contract test for this edge case.

File: packages/core/src/loop/lifecycle-state.mjs [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+    "ready_to_rerequest_review",
+  ]),
+  [LIFECYCLE_STATE.PRE_APPROVAL_GATE]: Object.freeze([
+    "low_signal_converged",
+    "round_cap_clean_fallback",
+    "internal_tooling_direct_gate",
+    "done",
+  ]),
+  [LIFECYCLE_STATE.MERGE]: Object.freeze([]),
+});
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/loop/lifecycle-state.mjs:326]: Fixed in fde57f7: COPILOT_INNER_STATE_MAP now exhaustive. Added waiting_for_ci, review_request_unavailable, blocked_needs_user_decision, round_cap_reached -> feedback_resolution. Moved done -> merge (terminal). Tests: 55 pass.

File: packages/core/test/lifecycle-state.test.mjs [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+  assert.equal(lifecyclePhaseForCopilotState("pr_ready_no_feedback"), LIFECYCLE_STATE.DRAFT_GATE);
+  assert.equal(lifecyclePhaseForCopilotState("waiting_for_copilot_review"), LIFECYCLE_STATE.FEEDBACK_RESOLUTION);
+  assert.equal(lifecyclePhaseForCopilotState("unresolved_feedback_present"), LIFECYCLE_STATE.FEEDBACK_RESOLUTION);
+  assert.equal(lifecyclePhaseForCopilotState("already_fixed_needs_reply_resolve"), LIFECYCLE_STATE.FEEDBACK_RESOLUTION);
+  assert.equal(lifecyclePhaseForCopilotState("ready_to_rerequest_review"), LIFECYCLE_STATE.FEEDBACK_RESOLUTION);
+  assert.equal(lifecyclePhaseForCopilotState("done"), LIFECYCLE_STATE.PRE_APPROVAL_GATE);
+  assert.equal(lifecyclePhaseForCopilotState("low_signal_converged"), LIFECYCLE_STATE.PRE_APPROVAL_GATE);
+  assert.equal(lifecyclePhaseForCopilotState("round_cap_clean_fallback"), LIFECYCLE_STATE.PRE_APPROVAL_GATE);
+  assert.equal(lifecyclePhaseForCopilotState("internal_tooling_direct_gate"), LIFECYCLE_STATE.PRE_APPROVAL_GATE);
+});
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/test/lifecycle-state.test.mjs:496]: Fixed in fde57f7: COPILOT_INNER_STATE_MAP now exhaustive. Added waiting_for_ci, review_request_unavailable, blocked_needs_user_decision, round_cap_reached -> feedback_resolution. Moved done -> merge (terminal). Tests: 55 pass.

File: packages/core/src/loop/lifecycle-state.mjs [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+// Connection to copilot-loop-state.mjs inner machine
+// ---------------------------------------------------------------------------
+
+/**
+ * Map from outer lifecycle phases to copilot-loop-state.mjs inner machine states.
+ *
+ * Skills use this mapping to determine which inner-machine states are active
+ * during a given lifecycle phase. Not all lifecycle phases have a corresponding
+ * inner-machine state (issue_intake, refinement, merge are outer-only).
+ *
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/loop/lifecycle-state.mjs:292]: Fixed in fde57f7: COPILOT_INNER_STATE_MAP now exhaustive. Added waiting_for_ci, review_request_unavailable, blocked_needs_user_decision, round_cap_reached -> feedback_resolution. Moved done -> merge (terminal). Tests: 55 pass.

File: packages/core/test/lifecycle-state.test.mjs [hunk 11]
```diff
@@ -0,0 +1,521 @@
...
+      `${phase} should have an inner state map entry`,
+    );
+  }
+});
+
+test("issue_intake and refinement and merge map to empty inner states (outer-only)", () => {
+  assert.deepEqual(COPILOT_INNER_STATE_MAP[LIFECYCLE_STATE.ISSUE_INTAKE], []);
+  assert.deepEqual(COPILOT_INNER_STATE_MAP[LIFECYCLE_STATE.REFINEMENT], []);
+  assert.deepEqual(COPILOT_INNER_STATE_MAP[LIFECYCLE_STATE.MERGE], []);
+});
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/test/lifecycle-state.test.mjs]: Fixed in fde57f7: COPILOT_INNER_STATE_MAP now exhaustive. Added waiting_for_ci, review_request_unavailable, blocked_needs_user_decision, round_cap_reached -> feedback_resolution. Moved done -> merge (terminal). Tests: 55 pass.

File: packages/core/test/lifecycle-state.test.mjs [hunk 12]
```diff
@@ -0,0 +1,521 @@
...
+
+test("feedback_resolution maps to review/fix inner states", () => {
+  const inner = COPILOT_INNER_STATE_MAP[LIFECYCLE_STATE.FEEDBACK_RESOLUTION];
+  assert.deepEqual([...inner].sort(), [
+    "waiting_for_copilot_review",
+    "unresolved_feedback_present",
+    "already_fixed_needs_reply_resolve",
+    "ready_to_rerequest_review",
+  ].sort());
+});
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/test/lifecycle-state.test.mjs]: Fixed in fde57f7: COPILOT_INNER_STATE_MAP now exhaustive. Added waiting_for_ci, review_request_unavailable, blocked_needs_user_decision, round_cap_reached -> feedback_resolution. Moved done -> merge (terminal). Tests: 55 pass.

File: packages/core/test/lifecycle-state.test.mjs [hunk 13]
```diff
@@ -0,0 +1,521 @@
...
+
+test("lifecyclePhaseForCopilotState returns null for unknown inner states", () => {
+  assert.equal(lifecyclePhaseForCopilotState("blocked_needs_user_decision"), null);
+  assert.equal(lifecyclePhaseForCopilotState("review_request_unavailable"), null);
+  assert.equal(lifecyclePhaseForCopilotState("round_cap_reached"), null);
+  assert.equal(lifecyclePhaseForCopilotState("unknown_state"), null);
+  assert.equal(lifecyclePhaseForCopilotState(""), null);
+  assert.equal(lifecyclePhaseForCopilotState(null), null);
+});
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/test/lifecycle-state.test.mjs]: Fixed in fde57f7: COPILOT_INNER_STATE_MAP now exhaustive. Added waiting_for_ci, review_request_unavailable, blocked_needs_user_decision, round_cap_reached -> feedback_resolution. Moved done -> merge (terminal). Tests: 55 pass.

File: packages/core/src/loop/lifecycle-state.mjs [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+// ---------------------------------------------------------------------------
+// Connection to copilot-loop-state.mjs inner machine
+// ---------------------------------------------------------------------------
+
+/**
+ * Map from outer lifecycle phases to copilot-loop-state.mjs inner machine states.
+ *
+ * Skills use this mapping to determine which inner-machine states are active
+ * during a given lifecycle phase. Not all lifecycle phases have a corresponding
+ * inner-machine state (issue_intake, refinement, merge are outer-only).
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/loop/lifecycle-state.mjs:291]: Fixed in 426a67f: updated docstring — issue_intake and refinement are outer-only; merge maps to done (terminal inner state).

**Comment by mfittko** [comment]
### Gate review: `pre_approval_gate`

**Reviewed head SHA:** `426a67fdda2e4ce4682280ab35e5082850b57b09`
**Verdict:** clean

**Findings summary:** Pre-approval gate: 2 files (lifecycle-state.mjs + test), 55/55 tests pass, CI green, 0 unresolved threads. 5 Copilot rounds with all findings addressed. Module: LIFECYCLE_STATE (7 phases), LIFECYCLE_TRANSITIONS, LIFECYCLE_GRAPH, resolveLifecycleState, exhaustive COPILOT_INNER_STATE_MAP (12 inner states), lifecyclePhaseForCopilotState. Contract tests cover state constants, transitions, resolver, and inner-state connectivity. No unrelated files. Scope matches #676.; Copilot review rounds exhausted (5/5); current head has zero unresolved threads and green or credibly green CI, so pre_approval_gate fallback is allowed without another Copilot re-request.

**Next action:** Ready for human approval; merge after authorization.
